### PR TITLE
[sw, dif_aes] Introduce a set of AES changes 

### PR DESCRIPTION
**Pull request by silvestrst**
The aim of this PR is to merge in the main library with other TODOs (**described in #4833 **) addressed in follow-ups.

Please have a look at the #2382 , for the questions and suggestions that I have gathered whilst working my way through the [DIF AES OpenTitan documentation](https://docs.opentitan.org/hw/ip/aes/doc/).

Key points:

- AES API is relatively straight forward, however there is a number of internal conditions that have to be met for correct operation of the peripheral. These are fundamental checks, and should be done at the DIF level rather than the driver level (in my opinion).

- It does not seem practical to bring the input or output data pointers into alignment as it is done in HMAC and SPI. The data that AES operates on is 128-bits wide, and it would be costly to bring it into the alignment for every 128-bit read/write. I think it should be documented that for maximal efficiency the input/output buffers have to be aligned.

### CODE REVIEW

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [inline sw/device/lib/dif/dif_aes.h:58]: I was itching to store these values in `dif_aes` after configuration (caching the current configuration for DIF internal use), to avoid a number of mmio operations. I see this more like in dif_spi where rx sizes are stored in `dif_aes`. However, for now I am checking the configuration values internally through mmio  operations. What do you think?

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:232]: Sure, but wouldn't it place additional requirements on the caller - forcing the buffers that they feed to the function to be of `uint32_t` type to prevent aliasing issues?

If it's `void*` and internally we use `memcpy`, then compiler should be able to figure out when the buffer is aligned and do the copy efficiently?

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:269]: Please see my question above.

**Comment by silvestrst** [comment]
> The bulk of the review boils down to replacing the individual behavior setters with combined start functions.

I was thinking about that as well, sounds good. I am planning to do some more AES work next week, so I will address the issues then - and also with that add test cases and implementation.

@lenary and @gkelly thank you for the review.

File: sw/device/lib/dif/dif_aes.h
Comment (lenary) [reply sw/device/lib/dif/dif_aes.h:232]: We decided it was better to reflect that AES is handling 4 32-bit integers, rather than to just have this parameter be a buffer. 

One of the major things we didn't like was that `void*` has an implicit length, something the struct makes much more explicit, which we think is a lot safer.

Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:232]: `char *example_message[19] = "123412341234123412" // 19 bytes with \0`
How would you propose to feed the handle `example_message` into `somefunc`, given:
`typedef struct somestruct { uint32_t someval[4]; } somestruct_t;`
`void somefunc(somestruct_t input_data);`

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:20]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:161]: I do think this is useful to keep. For example in `ECB` mode as far as I understand we should not use the same key for more than one block of data. I think the intended use-case is to change the key for every block of input data in ECB mode, then setting up a number of register fields that don't need to be changed is wasteful.

Comment (vogelpi) [reply sw/device/lib/dif/dif_aes.h:161]: Actually, ECB should not be used at all for the reason mentioned by @silvestrst .

Comment (lenary) [reply sw/device/lib/dif/dif_aes.h:161]: You would only be using ECB mode if you were encrypting/decrypting a single block, right? Then we may want to slightly modify the ECB's start function, but this function has to be deleted, as it doesn't fit with the transactional nature of the API we requested from the design review.

If we don't want to expose ECB at all through the DIF that's a different issue.

Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:161]: > You would only be using ECB mode if you were encrypting/decrypting a single block, right?

I am not sure that the use-cases are limited to a single block, but maybe they are, I am no expert.

> Then we may want to slightly modify the ECB's start function, but this function has to be deleted, as it doesn't fit with the transactional nature of the API we requested from the design review.

There could be things that you did not consider in the design review, and hence this thread.

> If we don't want to expose ECB at all through the DIF that's a different issue.

It seems that it aligns with what @vogelpi said, however, if it should not be used at all, should we eliminate it from the peripheral completely?

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:123]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:194]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:231]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:268]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:269]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:232]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:161]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:110]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:59]: I decided to use this structure as a parameter for `dif_aes_<mode>_start` functions "- cipher_mode". It is not used in `init` anymore.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:33]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:303]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:353]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (vogelpi) [reply sw/device/lib/dif/dif_aes.h:161]: Sorry to bother you again. I am not fully familiar with the DIFs but I just wanted to confirm that really higher-level software should not use ECB at all, but that we need it at least for DV purposes. If there is a DV for ECB, it should be documented that is should not be used other than for DV.

Having the new ECB-specific function sounds reasonable.

Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:161]: > Sorry to bother you again.

No bother at all, in fact thank you for your help!

Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h]: > If there is a DV for ECB, it should be documented that is should not be used other than for DV.

I put a note saying that it is discouraged to use this mode, but maybe it needs stronger wording.

**Comment by silvestrst** [comment]
I have addressed all of the review comments, and have completely refactored the API as suggested. The implementation is completed as well, however has not been tested yet. Some of the implementation details probably could be further improved, like for example: CBC and CTR start functions can be probably internally split out into a common static function. However, I think having separate public API for these functions is a correct choice.

I think this PR needs at least one more high level header pass, and if there are not too many interface changes - I will promote this PR from "draft" to "open", and add other reviewers.

**Comment by silvestrst** [comment]
@vogelpi Thank you for your review, much appreciated. I will wait until @lenary and @gkelly have a chance to run through this PR again, and will fix up all outstanding comments in one go.

File: sw/device/lib/dif/dif_aes.h
Comment (lenary) [reply sw/device/lib/dif/dif_aes.h:76]: Yeah, this is a complex one. I'm not sure how we should solve this, but I agree it is confusing. We might just have to use enable/disable and communicate this with the parameter name. I'm not sure `gear` or `transmission` get the right points across - maybe something like `algorithm_advance` or something? (where the options are `auto` or `manual`)

**Comment by lenary** [comment]
ECB mode changes can come in a follow-up PR (as should any other mode additions), I feel. Let's see if any of pirmin's comments should also go into a follow-up, or if they're small enough to fix quickly.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h]: Done.

**Review by silvestrst — COMMENTED** [review]
I have addressed all of the comments apart from (which I intend to resolve in a follow-up):
> The DIF currently provides two separate functions for providing new data (load) and reading the output data. These functions check if new data can be loaded or if new output data is available. This is how these generic functions should work. However, when going for maximum performance, the input ready bit does not need to be checked anymore after the second block as illustrated in the block operation section in the technical spec. This allows you to spare 1 out of 10 peripheral accesses which will probably be significant when processing large messages. I think the DIF should provide either a function for doing large transfers directly, or a separate load function that does not check the input ready bit (with a note that drivers should use this one to implement block operation as mentioned in the spec). Maybe this collides a bit with other requirements/guidelines on DIFs and we might need to find another way. We can also discuss this.

Additionally, I would need to have a look at it closely - as if the speed considerations are so important there might be other boosters apart from an `if` check. At a glance - this `if` alone is not a major performance booster.

**Comment by silvestrst** [comment]
Still need to comment the data types, and refactor to match the DIF template that we have.

File: sw/device/lib/dif/dif_aes.h
Comment (vogelpi) [inline sw/device/lib/dif/dif_aes.h:298]: @paran -> @param?

**Review by vogelpi — COMMENTED** [review]
Quite some time has passed since starting this PR, and in the meantime quite some changes have been done in the hardware. Some more will land in a couple of days. I have thus done a full review and added pointers to relevant PRs. Sorry for the many comments.

Three main things to consider. We don't need to do this in this PR but at least for follow up PRs:
1. We now have 2 more block cipher modes of operation: CFB-128 and OFB.
2. The performance thing mentioned in previous comments: it's not just about doing an `if`. The `if` does an additional register read from the peripheral over TL-UL. Unlike for other devices we have, AES operation is very quick: you get a new 128-bit block after 12 to 16 clock cycles. Saving even a single bus transaction can be significant.
3. ECB (sorry, we have been discussing a lot already about this): It turns out that for SCA, we primarily use ECB because SCA is mainly about leakage in the cipher core. So you don't want the other modes to obfuscate things. Re-Writing the key for every block is impractical and can become a performance bottleneck, especially as we're already in the millions of traces needed for analysis.

But as I said, we don't need to change this now. I think we shouldn't block this PR further and move on.

**Comment by silvestrst** [comment]
Hey @vogelpi , thanks for the review. Also there is this #4833. The additional modes implementation will go to @alphan as a part of handover. I will see how many of the above comments it makes sense to fix-up myself and how many leave for Alphan as part of handover. I am planning on going through the comments today.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:499]: Thanks.

Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:538]: One of us (me or Alphan will add it, when PR lands).

File: sw/device/lib/dif/dif_aes.h
Comment (mcy) [inline sw/device/lib/dif/dif_aes.h:570]: Hmmm, `get_status` for consistency with other DIFs?

File: sw/device/lib/dif/dif_aes.h
Comment (mcy) [inline sw/device/lib/dif/dif_aes.h:567]: There seems to be a mix of `dif_aes_result_t` and "The result of the operation" values for this parameter. Maybe use one and be consistent? Same for the `aes` param.

File: sw/device/lib/dif/dif_aes.c
Comment (mcy) [inline sw/device/lib/dif/dif_aes.c:25]: What does this TODO mean?

File: sw/device/lib/dif/dif_aes.c
Comment (mcy) [inline sw/device/lib/dif/dif_aes.c:84]: This semicolon is redundant.

File: sw/device/lib/dif/dif_aes.c
Comment (mcy) [inline sw/device/lib/dif/dif_aes.c:136]: File a bug...?

File: sw/device/lib/dif/dif_aes.c
Comment (mcy) [inline sw/device/lib/dif/dif_aes.c:143]: Should this be a stanza?

File: sw/device/lib/dif/dif_aes.c
Comment (mcy) [inline sw/device/lib/dif/dif_aes.c:146]: Why not inline this condition into the `true` below?

File: sw/device/lib/dif/dif_aes.c
Comment (mcy) [inline sw/device/lib/dif/dif_aes.c:147]: This `0` seems to imply that the previous manipulations to `reg` are discarded. Is this a bug?

File: sw/device/lib/dif/dif_aes.c
Comment (mcy) [inline sw/device/lib/dif/dif_aes.c:370]: For a followup, it seems like this wants to return a bitset...?

File: sw/device/lib/dif/dif_aes.c
Comment (mcy) [inline sw/device/lib/dif/dif_aes.c:408]: Nit: bitfield_bit_index_t.

File: sw/device/lib/dif/dif_aes.h
Comment (mcy) [inline sw/device/lib/dif/dif_aes.h:19]: Drop this comment.

File: sw/device/lib/dif/dif_aes.c
Comment (vogelpi) [reply sw/device/lib/dif/dif_aes.c:159]: Thanks :+1:

File: sw/device/lib/dif/dif_aes.c
Comment (vogelpi) [inline sw/device/lib/dif/dif_aes.c:21]: It seems you have accidentally removed the comment explaining the values and resolution of invalid values. Do you want to add that back?

File: sw/device/lib/dif/dif_aes.h
Comment (vogelpi) [inline sw/device/lib/dif/dif_aes.h:61]: I think you wanted to change this to `share0` similarly `share1` below. Did you maybe forget to commit/push?

File: sw/device/lib/dif/dif_aes.h
Comment (vogelpi) [inline sw/device/lib/dif/dif_aes.h:167]: Really sorry to bother you again about this. I had to rename it again in #4909 . It's now `kDifAlertFatalFault` and `kDifAlertRecovCtrlUpdateErr`.

File: sw/device/lib/dif/dif_aes.h
Comment (vogelpi) [inline sw/device/lib/dif/dif_aes.h:554]: Also these have changed again in line with the alert names. Sorry for that.

File: sw/device/lib/dif/dif_aes.h
Comment (vogelpi) [reply sw/device/lib/dif/dif_aes.h:540]: This got merged now.

File: sw/device/lib/dif/dif_aes.h
Comment (vogelpi) [reply sw/device/lib/dif/dif_aes.h]: it's still @paran. It seems that some changes didn't get committed/pushed yet.

File: sw/device/lib/dif/dif_aes.c
Comment (vogelpi) [reply sw/device/lib/dif/dif_aes.c:25]: It means these modes still need to be implemented. This is tracked in #4833.

**Review by vogelpi — COMMENTED** [review]
Sorry for keeping you waiting @silvestrst . This looks mostly good but I think you forgot to commit/push some changes you've made (as you commented on them).

If the remaining things and also @mcy comments are addressed (I think all these are quick things) I am happy to approve this PR and move forward.

File: sw/device/lib/dif/dif_aes.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.c]: Done.

File: sw/device/lib/dif/dif_aes.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.c]: Done.

File: sw/device/lib/dif/dif_aes.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.c]: > File a bug...?

You mean an issue?

Anyhow - this has been removed.

File: sw/device/lib/dif/dif_aes.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.c]: Not sure what you mean, I split up all `bitfield_field32_write` calls so that they are all on their own.

File: sw/device/lib/dif/dif_aes.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.c:151]: I think it is fine either way, keeping as is.

File: sw/device/lib/dif/dif_aes.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.c]: This is a bug, and If I saw this comment earlier, it would have saved me some time in debugging :)

File: sw/device/lib/dif/dif_aes.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.c]: Possibly, I will leave it with @alphan to decide.

File: sw/device/lib/dif/dif_aes.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.c]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h]: Possibly, thanks.

Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h:539]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h]: Done.

File: sw/device/lib/dif/dif_aes.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_aes.h]: Done.

**Review by silvestrst — COMMENTED** [review]
Thanks again guys for your review, I should have addressed all of the outstanding comments now.

I think we should try to merge this now unless there are major issues.

**Review by vogelpi — APPROVED** [review]
Thanks @silvestrst , this looks good to me!

To actually merge the PR, I think also @gkelly will need to approve it as he requested changes a long time ago. The interface still says: "Changes requested".

**Review by gkelly — APPROVED** [review]

**Pull request merged**
